src/acceptor.rs

```diff
@@ -0,0 +1,535 @@
+                Err(e) => map_error(e),
+                Ok(Some(accepted)) => {
+                    BoxCastPtr::set_mut_ptr(out_accepted, Some(accepted));
+                    rustls_result::Ok
+                }
+            }
+        }
+    }
+}
+
+impl rustls_accepted {
+    /// Return the server name indication (SNI) from the ClientHello.
+    /// If the SNI contains a NUL byte, return a zero-length rustls_str.
+    /// Also return a zero-length rustls_str if there was some other
+    /// usage error, like calling with a NULL pointer or with an already-used
+    /// rustls_accepted. The returned rustls_str is valid until the next
+    /// time a method is called on the `rustls_accepted`. It is not owned
+    /// by the caller and does not need to be freed.
+    #[no_mangle]
+    pub extern "C" fn rustls_accepted_server_name(
+        accepted: *const rustls_accepted,
+    ) -> rustls_str<'static> {
+        ffi_panic_boundary! {
+            let accepted: &Option<Accepted> = try_ref_from_ptr!(accepted);
+            let accepted = match accepted {
+                Some(a) => a,
+                None => return Default::default(),
+            };
+            let hello = accepted.client_hello();
+            let sni = match hello.server_name() {
+                Some(s) => s,
+                None => return Default::default(),
+            };
+            match rustls_str::try_from(sni) {
+                Ok(s) => unsafe { s.into_static() },
+                Err(_) => Default::default(),
+            }
+        }
+    }
+
+    /// Return the i'th in the list of signature schemes offered in the ClientHello.
+    /// This is useful in selecting a server certificate when there are multiple
+    /// available for the same server name, for instance when selecting
+    /// between an RSA and an ECDSA certificate. Returns 0 if i is past the end of
+    /// the list or on a usage error, like calling with a NULL pointer or an
+    /// already-used rustls_accepted.
+    #[no_mangle]
+    pub extern "C" fn rustls_accepted_signature_scheme(
+        accepted: *const rustls_accepted,
+        i: usize,
+    ) -> u16 {
+        ffi_panic_boundary! {
+            let accepted: &Option<Accepted> = try_ref_from_ptr!(accepted);
+            let accepted = match accepted {
+                Some(a) => a,
+                None => return 0,
+            };
+            let hello = accepted.client_hello();
+            let signature_schemes = hello.signature_schemes();
+            if i < signature_schemes.len() {
+                signature_schemes[i].get_u16()
+            } else {
+                0
+            }
+        }
+    }
+
+    /// Return the i'th ALPN protocol requested by the client in the ClientHello.
+    /// If the client did not offer the ALPN extension, return a zero-length rustls_slice_bytes.
+    /// Also return a zero-length rustls_slice_bytes on a usage error, like
+    /// calling with a NULL pointer or an already-used rustls_accepted.
```

Comment:
I noticed that the comment for `rustls_accepted_signature_scheme()` calls out the case where `i`is past the end of the slice, whereas this doesn't (although it talks generically about "usage error"). Maybe good to make that explicit?

In general, I feel like these comments could do with more structuring, like maybe different paragraphs about arguments and invariants and return values? The blob of text makes it harder to find the relevant concerns (though maybe this is specific to the way I'm reviewing here).

Reply:
Good idea. I updated all the comments to have `Parameters`, `Returns`, and `Memory and lifetimes` sections. I think it makes things a lot clearer. I'll update the other comments in the library to match the format after this lands.

Reply:
I feel like it might be more clear if the relevant memory and lifetimes concerns were attached to the particular parameters and return value?

Reply:
I moved some of the explanations inline with the parameters. I left a few that felt like they were too long to be described inline.